gossip/src/duplicate_shred_handler.rs

```diff
@@ -0,0 +1,431 @@
+        self.chunk_map.insert(slot, SlotStatus::Frozen);
+        for (_, current_slots_set) in self.validator_pending_proof_map.iter_mut() {
+            current_slots_set.remove(&slot);
+        }
+    }
+
+    fn insert_chunk(&mut self, data: DuplicateShred) -> Result<Option<Vec<DuplicateShred>>, Error> {
+        if let SlotStatus::UnfinishedProof(slot_chunk_map) = self
+            .chunk_map
+            .entry(data.slot)
+            .or_insert_with(|| SlotStatus::UnfinishedProof(HashMap::new()))
+        {
+            let proof_chunk_map = slot_chunk_map
+                .entry(data.from)
+                .or_insert_with(|| Self::new_proof_chunk_map(data.num_chunks, data.wallclock));
+            if proof_chunk_map.wallclock < data.wallclock {
+                proof_chunk_map.num_chunks = data.num_chunks;
+                proof_chunk_map.missing_chunks = data.num_chunks;
+                proof_chunk_map.wallclock = data.wallclock;
+                proof_chunk_map.chunks.clear();
+            }
+            let num_chunks = data.num_chunks;
+            let chunk_index = data.chunk_index;
+            let slot = data.slot;
+            let from = data.from;
+            if num_chunks == proof_chunk_map.num_chunks
+                && chunk_index < num_chunks
+                && !proof_chunk_map.chunks.contains_key(&chunk_index)
+            {
+                proof_chunk_map.missing_chunks = proof_chunk_map.missing_chunks.saturating_sub(1);
+                proof_chunk_map.chunks.insert(chunk_index, data);
+                if proof_chunk_map.missing_chunks == 0 {
+                    let mut result: Vec<DuplicateShred> = Vec::new();
+                    for i in 0..num_chunks {
+                        result.push(proof_chunk_map.chunks.remove(&i).unwrap())
+                    }
+                    return Ok(Some(result));
+                }
+            }
+            self.validator_pending_proof_map
+                .entry(from)
+                .or_insert_with(HashSet::new)
+                .insert(slot);
+        }
+        Ok(None)
+    }
+
+    fn verify_and_apply_proof(&self, slot: Slot, chunks: Vec<DuplicateShred>) -> Result<(), Error> {
+        if slot <= self.blockstore.last_root() || self.blockstore.has_duplicate_shreds_in_slot(slot)
+        {
+            return Ok(());
+        }
+        let (shred1, shred2) = into_shreds(chunks, |slot| {
+            self.leader_schedule_cache.slot_leader_at(slot, None)
+        })?;
+        self.blockstore
+            .store_duplicate_slot(slot, shred1.into_payload(), shred2.into_payload())?;
+        Ok(())
+    }
+
+    fn cleanup_old_slots(&mut self) {
+        let new_last_root = self.blockstore.last_root();
+        if self.last_root < new_last_root {
+            self.chunk_map.retain(|k, _| k > &new_last_root);
+            for (_, slots_sets) in self.validator_pending_proof_map.iter_mut() {
+                slots_sets.retain(|k| k > &new_last_root);
+            }
```

Comment:
If we want to discard all keys below some threshold it is probably better to use `BTreeMap` and `BTreeSet`, and use `split_off` instead. Similar to:
https://github.com/solana-labs/solana/blob/055c6a5eb/core/src/cluster_slots.rs#L96

Reply:
does this still need to be addressed?

Reply:
Not yet, when we switch to a different method of deciding whether pubkeys are valid we might need to totally rework the data structure. Keeping as is to make this change small.